### PR TITLE
Jh/reorganize

### PR DESCRIPTION
**Pull request by joelhooks**
This reorganizes this component considerably with an aggressive use of render props. I also chose to use downshift 🏎  for the selector list for now, because it already exists and was easy to drop in. It is going to be removed as a dependency because this isn't really a drop-down type list and downshift doesn't quite fit/make sense.

The selected list uses react-beautiful-dnd which will definitely be used as a dependency. Writing a drag-n-drop library isn't something I care to explore ;)

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,22 +1,229 @@
...
...cgYW5kIGNvbmZpZ3VyaW5nIEFuZ3VsYXIgTWF0ZXJpYWwgdG8gYmUgdXNlZCBpbiBhIHByb2plY3QuIFRoaXMgd2lsbCBpbmNsdWRlIGNvbmZpZ3VyaW5nIGEgY3VzdG9tIEFuZ3VsYXIgTWF0ZXJpYWwgbW9kdWxlIGFzIHdlbGwgYXMgc29tZSBwcmVidWlsdCB0aGVtZXMgYXZhaWxhYmxlIHRvIHVzLlxuV2UnbGwgbG9vayB0aHJvdWdoIHRoZSBNYXRlcmlhbCBhcGkgYXMgd2VsbCBhcyBhcHBseSB0aXBzIGFuZCB0cmlja3Mgb24gY29tbW9uIHByb2JsZW1zIHlvdSdsbCBmYWNlIHdoZW4gYXBwbHlpbmcgdGhpcyBzdHlsZSB0byB5b3VyIGFwcGxpY2F0aW9uLiBXZSdsbCBjb3ZlciB0b3BpY3Mgc3VjaCBhcyBob3cgTWF0ZXJpYWwgaGFuZGxlcyB1c2VyIGlucHV0LCBjb21tb24gTWF0ZXJpYWwgdXRpbGl0eSBjb21wb25lbnRzIChzcGlubmVycywgcHJvZ3Jlc3MgYmFycyBldGMuKSwgYW5kIGhvdyB0byBwcm9wZXJseSBtYW5hZ2UgdGFidWxhciBkYXRhIGFuZCB0aGVtZXMgaW4gQW5ndWxhciBNYXRlcmlhbC5cbkJ5IHRoZSBlbmQgb2YgdGhpcyBjb3Vyc2UsIHlvdSB3aWxsIGJlIGNvbWZvcnRhYmxlIGxldmVyYWdpbmcgdGhlIEFuZ3VsYXIgTWF0ZXJpYWwgY29tcG9uZW50IGxpYnJhcnkgdG8gY3JlYXRlIGFwcGxpY2F0aW9ucyB0aGF0IGxvb2sgYW5kIGZlZWwgZ3JlYXQgYmFzZWQgb24gdGhlIE1hdGVyaWFsIGRlc2lnbiBndWlkZWxpbmVzIGFuZCBoYXZlIGEgaGlnaCBmb2N1cyBvbiBhY2Nlc3NpYmlsaXR5LlxuIiwiY3VycmVudF9jb3Vyc2VfbmV4dF9sZXNzb25fdXJsIjoiaHR0cHM6Ly9lZ2doZWFkLmlvL2xlc3NvbnMvYW5ndWxhci1hbmd1bGFyLW1hdGVyaWFsLWNvbXBvbmVudHMtY291cnNlLW92ZXJ2aWV3In19.5dJgAm43vwS5Xh6I-D-kpY_b_J-kREI6srMMj5WJt5M';
```
Comment (joelhooks) [inline src/App.js:12]: http://egghead.af:5000/users/jwt?return_to=http://localhost:3000 will give you a token

File: src/App.js [hunk 2]
```diff
@@ -1,22 +1,229 @@
...
+  state = {
+    selectedLessons,
+    allLessons,
+    isOpen: false
+  }
+
+  handleStateChange = debounce((changes) => {
+    if (changes.hasOwnProperty('inputValue') && !isUndefined(changes['inputValue'])) {
+      return http
+        .get(`http://egghead.af:5000/api/v1/lessons?q=${changes['inputValue']}&size=20&page=1`)
```
Comment (joelhooks) [inline src/App.js:26]: this requires a [local elasticsearch setup](https://gist.github.com/jpalala/ab3c33dd9ee5a6efbdae). Once that is installed you need to `Lesson.reindex` inside of the rails console (`rails c`). Also assumes this PR is merged: https://github.com/eggheadio/egghead-rails/pull/2228

File: src/App.js [hunk 3]
```diff
@@ -1,22 +1,229 @@
...
+      fn && fn(event, ...args)
+      return event.defaultPrevented
+    })
+}
+
+function itemToString(item) {
+  return item.slug
+}
+
+class List extends React.Component {
```
Comment (joelhooks) [inline src/App.js:199]: now that we have a working component above with no style applied at all, but mostly feature complete, the next step will be to clean it all up and create a "ContentSelector" component that fully encapsulates the behavior here and can be used and styled easily to suit different content selection needs across the app.

File: src/App.js [hunk 4]
```diff
@@ -1,22 +1,229 @@
...
+    return {
+      onClick: composeEventHandlers(onClick, () => {
+        console.log('click click')
+      }),
+      ...rest,
+    }
+  }
+
+  render() {
+    const children = unwrapArray(this.props.render || this.props.children, noop)
```
Comment (theianjones) [inline src/App.js:224]: What is the purpose of passing the noop?

File: src/App.js [hunk 5]
```diff
@@ -1,22 +1,229 @@
...
+    this.setState({
+      selectedLessons: remove(this.state.selectedLessons, (itemToRemove) => !isEqual(itemToRemove, item))
+    })
+  }
+
+  selectItem = (item) => {
+    if(some(this.state.selectedLessons, item)) {
+      this.removeItem(item)
+    } else {
+      this.setState({
```
Comment (theianjones) [inline src/App.js:42]: This else handles adding an item to the list?

File: src/App.js [hunk 6]
```diff
@@ -1,22 +1,229 @@
...
+    }
+
+  }
+
+  onDragStart = (initial) => {
+    console.log(initial)
+    // Add a little vibration if the browser supports it.
+    // Add's a nice little physical feedback
+    if (window.navigator.vibrate) {
+      window.navigator.vibrate(100);
```
Comment (theianjones) [inline src/App.js:55]: I didn't know about this. Cool

**Review by theianjones — APPROVED** [review]
Render props seem to make this quite nice.

File: src/App.js [hunk 7]
```diff
@@ -1,22 +1,229 @@
...
+      getItemProps,
+      selectItem
+    }
+  }
+
+  selectItem = () => {
+    console.log('do select')
+  }
+
+  getItemProps = ({onClick, ...rest}) => {
```
Comment (AkashGutha) [inline src/App.js:214]: not a big fan of this method. IMO it makes it harder to understand. 
I'll look into this to find an alternate solution that doesn't hijack the code and provides the same functionality.

File: src/App.js (hunk 5 above)
Comment (joelhooks) [reply src/App.js:42]: just being lazy, it's more of a "toggle item selection" with this conditional in here.

File: src/App.js (hunk 4 above)
Comment (joelhooks) [reply src/App.js:224]: to avoid `children is not a function` error

File: src/App.js (hunk 7 above)
Comment (joelhooks) [reply src/App.js:214]: this approach allows me the user to pass in arbitrary properties. So our known/required properties are captured and utilized, but the consumer has the ultimate flexibility to affect change on the item element as required for their unknown (to us), specific use case.

it's a pattern for highly customizable reusable components and isn't as difficult to understand once it is recognized.

You could elaborate on "hijack the code". Always interesting to hear the other side of this specific debate.

**Pull request merged**